README.md
reflect-copy
===================
A verified copy for Node.js, available across filesystems and networks. It even works for directories.

Master Build Status: 
[![Build Status](https://travis-ci.org/aerisweather/node-reflect-copy.svg?branch=master)](https://travis-ci.org/aerisweather/node-reflect-copy)
[![Coverage Status](https://coveralls.io/repos/aerisweather/node-reflect-copy/badge.svg?branch=master&service=github)](https://coveralls.io/github/aerisweather/node-reflect-copy?branch=master)

__This project is sponsored by:__

[![AerisWeather](http://branding.aerisweather.com/logo-dark-small.png)](http://www.aerisweather.com) - Empowering the next generation, [aerisweather.com](https://www.aerisweather.com)

Our use case deals with a mounted remote file system. We need a way to reliably copy things across the network. Since the remote 
is just a folder, we have limited capabilities, so we read in the file(s), get a hash of them, copy them over, then verify our copy
produces the same hash. Not fool proof, but pretty good at ensuring there weren't network hiccups along the way.

## Quick Example

```javascript
var rCopy = require('reflect-copy').copy;

rCopy('/path/to/source', 'path/to/dest', {}, function(err, results) {
    console.log(results.source); //The source file's hash
    console.log(results.destination); //The destination file's hash
});
```
## Features
We needed a nice copy to clone files across file systems. File operations should be verifiable yet simple. Using checksums we can ensure a file has been copied correctly, in it's entirety. We use the module `fs-extra` (because it's awesome) as a base, then just wrap it with a nice checksum verification method.

* Pre and post checksum comparison
* Locking for file copies
* Copy to temp location first, move into position (useful for network transfers)
* Configuration for common options

## Options

| Option | Type | Default | Description |
| ------ | ---- | ------- | ----------- |
| `clobber` | (boolean) | `false` | Whether to overwrite if this file exists. |
| `srcHash` | (string) | `(empty)` | The known hash of the source file, if known |
| `hashMethod` | (string) | `'md5'` | Which hash to use for the checksum: 'md5' or 'sha1' |
| `tmpSuffix` | (string) | `'.copy-tmp'` | Whether to overwrite if this file exists. |
| `useLock` | (boolean) | `true` | (Locking) Should we use lock files to ensure we are the only reflect-copy writing to this location. |
| `stale` | (int, ms) | `30000` | (Locking) Duration to wait for the lock file before it is deemed stale (in milliseconds) |
| `wait` | (int, ms) | `10000` | (Locking) How long to wait for a lock file before erroring (in milliseconds) |

## Roadmap

 1. Support folders with multiple files
 1. Create a copySync method


CHANGELOG.md
Changelog
=========

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased full version

This project is currently not in full release and is subject to change. (Still 0.x versions)

## [0.1.0] - 2015-08-23
### Changed
Initial options specified, all tests passing.


test/copy-test.js
var assert = require('assert'),
	fsExtra = require('fs-extra'),
	path = require('path'),
	reflectCopy = require('../index');

var fixturesDir = path.join(__dirname, 'fixtures');
var multipleDir = path.join(__dirname, 'fixtures', 'my-dir');
var resultDir = path.join(__dirname, 'dest');
describe('copy', function() {

	beforeEach(function() {
		// runs before each test in this block
		fsExtra.removeSync(resultDir);
	});

    afterEach(function() {
        // runs before each test in this block
        fsExtra.removeSync(resultDir);
    });

	it('should copy the file to an existing folder', function(done) {
		fsExtra.ensureDirSync(resultDir);
		reflectCopy.copy(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'), function(err, result) {
			if(err) {
				assert.ok(false, err.message);
			}
			fileExistsSync(path.join(resultDir, 'loon.jpg'));
			done();
		})
	});

    it('should copy the file using sha1 as a checksum', function(done) {
        fsExtra.ensureDirSync(resultDir);
        reflectCopy.copy(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'), {hashMethod: 'sha1'}, function(err, result) {
            if(err) {
                assert.ok(false, err.message);
            }
            fileExistsSync(path.join(resultDir, 'loon.jpg'));
            done();
        })
    });

    it('should error because bad hash specified', function(done) {
        fsExtra.ensureDirSync(resultDir);
        reflectCopy.copy(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'), {hashMethod: 'nope'}, function(err, result) {
            if(err) {
                assert.equal(err.code, 'BADHASH');
                return done();
            }
            assert.ok(false, "Didn't throw DIFF error when using a known hash.");
        })
    });

	it('should copy the file to an existing folder, no filename', function(done) {
		fsExtra.ensureDirSync(resultDir);
		reflectCopy.copy(path.join(fixturesDir, 'loon.jpg'), resultDir, function(err, result) {
			if(err) {
				assert.ok(false, err.message);
			}
			fileExistsSync(path.join(resultDir, 'loon.jpg'));
			done();
		})
	});

	it('should copy the file to a non-existent folder', function(done) {
		reflectCopy.copy(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'), function(err, result) {
			if(err) {
				assert.ok(false, err.message);
			}
			fileExistsSync(path.join(resultDir, 'loon.jpg'));
			done();
		});
	});

	it('should NOT copy over an existing file (clobber off)', function(done) {
		fsExtra.copySync(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'));

		reflectCopy.copy(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'), function(err, result) {
			if(err) {
				assert.equal(err.code, 'EEXIST');
				return done();
			}
			assert.ok(false, "Didn't throw EEXIST error when trying to clobber a file!");
		});
	});

	it('should copy over an existing file (clobber on)', function(done) {
		fsExtra.copySync(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'));

		reflectCopy.copy(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'), {clobber: true}, function(err, result) {
			if(err) {
				assert.ok(false, err.message);
			}
			fileExistsSync(path.join(resultDir, 'loon.jpg'));
			done();
		});
	});

	it('should copy the file to a non-existent folder without a lock', function(done) {
		reflectCopy.copy(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'), {useLock: false}, function(err, result) {
			if(err) {
				assert.ok(false, err.message);
			}
			fileExistsSync(path.join(resultDir, 'loon.jpg'));
			done();
		});
	});

    it('should copy use a known hash and compare after copy', function(done) {
        reflectCopy.copy(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'), {srcHash: "a6921e11627e0c995546086e7865eacb"}, function(err, result) {
            if(err) {
                assert.ok(false, err.message);
            }
            fileExistsSync(path.join(resultDir, 'loon.jpg'));
            done();
        });
    });

    it('should copy use a known hash and compare after copy and fail', function(done) {
        reflectCopy.copy(path.join(fixturesDir, 'loon.jpg'), path.join(resultDir, 'loon.jpg'), {srcHash: "abc123"}, function(err, result) {
            if(err) {
                assert.equal(err.code, 'DIFF');
                return done();
            }
            assert.ok(false, "Didn't throw DIFF error when using a known hash.");
        });
    });

    it('should copy a directory to an existing folder', function(done) {
        fsExtra.ensureDirSync(resultDir);
        reflectCopy.copy(multipleDir, resultDir, function(err, result) {
            if(err) {
                assert.ok(false, err.message);
            }
            fileExistsSync(path.join(resultDir, 'my-dir', 'loon.jpg'));
            fileExistsSync(path.join(resultDir, 'my-dir', 'world-test.txt'));
            done();
        });
    });

    it('should copy a directory to a new folder', function(done) {
        reflectCopy.copy(multipleDir, resultDir, function(err, result) {
            if(err) {
                assert.ok(false, err.message);
            }
            fileExistsSync(path.join(resultDir, 'my-dir', 'loon.jpg'));
            fileExistsSync(path.join(resultDir, 'my-dir', 'world-test.txt'));
            done();
        });
    });
});

function fileExistsSync(filePath) {
	// Query the entry
	var stats = fsExtra.lstatSync(filePath);
	assert.ok(!stats.isDirectory() && stats.isFile(), "File " + filePath + " does not exist");
}

index.js
var async = require('async'),
    crypto = require('crypto'),
    fsExtra = require('fs-extra'),
    lockFile = require('lockfile'),
    path = require('path'),
    tar = require('tar-fs');

var MODE_FILE = 'file';
var MODE_DIRECTORY = 'directory';
function copy(source, destination, options, callback) {
    //Parse Args
    if (typeof options === 'function' && !callback) {
        callback = options;
        options = {}
    } else if (typeof options === 'function' || options instanceof RegExp) {
        //options = {filter: options}
        callback(new Error('This form of copy does not support multiple files yet...'));
    }
    options = getCopyOptions(options);
    callback = callback || function () {
    };

    options.source = source;
    options.destination = destination;

    var mode;
    async.waterfall([
            function (callback) {
                //Look at the source, are we moving a file or a directory?
                fsExtra.lstat(source, function (err, stats) {
                    if (stats.isDirectory()) {
                        mode = MODE_DIRECTORY;
                        callback(null);
                    }
                    else if (stats.isFile()) {
                        mode = MODE_FILE;
                        callback(null);
                    }
                    else {
                        var sourceNotFound = new Error("Source file doesn't exist");
                        sourceNotFound.code = "EEXISTSRC"
                        callback(new Error("Source file doesn't exist"))
                    }
                });
            },
            function (callback) {
                //Check destination
                fsExtra.lstat(destination, function (err, stats) {
                    //[Keep this here, we need to make a decision based on an error right here.]
                    var dirName = destination;
                    if (mode === MODE_FILE) {
                        dirName = path.dirname(destination);
                    }
                    if (err) {
                        //Destination doesn't exist.
                        fsExtra.ensureDir(dirName, function (err) {
                            if(mode === MODE_DIRECTORY) {
                                var baseName = path.basename(source);
                                options.destination = path.join(destination, baseName);
                            }
                            else {
                                options.destination = destination;
                            }
                            return callback(err);
                        });
                    }
                    else {
                        if (stats.isDirectory()) {
                            var baseName = path.basename(source);
                            options.destination = path.join(destination, baseName);
                        }
                        else {
                            options.destination = destination;
                        }
                        return callback(null);
                    }
                });
            },
            function (callback) {
                //Lock file for copying, make sure no one else interrupts us.
                if (options.useLock) {
                    var lockFilePath = options.destination + '.copy-lock';
                    var lockFileOpts = {
                        wait: options.wait,
                        stale: options.stale
                    };
                    lockFile.lock(lockFilePath, lockFileOpts, function (err) {
                        if (err) {
                            return callback(err);
                        }
                        return callback(null);
                    });
                }
                else {
                    return callback(null);
                }
            },
            function (callback) {
                //Get the source hash
                if (!options.srcHash) {
                    if (mode === MODE_FILE) {
                        options.source = source;
                        getFileHash(source, options.hashMethod, callback);
                    }
                    else if (mode === MODE_DIRECTORY) {
                        options.source = source + ".tar";
                        options.destination += ".tar";
                        tar.pack(source).pipe(fsExtra.createWriteStream(options.source))
                            .on('finish', function () {
                                getFileHash(options.source, options.hashMethod, callback);
                            }
                        );

                    }
                }
                else {
                    callback(null, options.srcHash);
                }
            },
            function (sourceFileHash, callback) {
                //Copy to temp location
                options.tmpDestination = options.destination;
                options.tmpDestination += options.tmpSuffix;
                fsExtra.copy(options.source, options.tmpDestination, options, function (err) {
                    if (err) {
                        return callback(err);
                    }
                    if(mode === MODE_DIRECTORY) {
                        fsExtra.remove(options.source, function(err) {
                            if(err) {
                                return callback(new Error("Couldn't remove temporary source (tar file generated when moving directories)"));
                            }
                            return callback(null, sourceFileHash);
                        })
                    }
                    else {
                        return callback(null, sourceFileHash);
                    }
                });
            },
            function (sourceFileHash, callback) {
                //Move into final position
                fsExtra.move(options.tmpDestination, options.destination, {clobber: options.clobber}, function (err) {
                    if (err) {
                        if (err.code === 'EEXIST') {
                            err.message = 'File already exists at ' + options.destination + ' (EEXIST) and the option \'clobber\' was set to false';
                        }
                        return callback(err);
                    }
                    return callback(null, sourceFileHash);
                });
            },
            function (sourceFileHash, callback) {
                //Verify hash
                getFileHash(options.destination, options.hashMethod, function (err, dstFileHash) {
                    if (err) {
                        return callback(err);
                    }
                    if (dstFileHash != sourceFileHash) {
                        var diffError = new Error("Copied file hashes don't match. Source: " + sourceFileHash + ", Destination: " + dstFileHash);
                        diffError.code = "DIFF";
                        return callback(diffError);
                    }

                    if (mode === MODE_DIRECTORY) {
                        var destWithoutTar = options.destination.substr(0, options.destination.length - 4);
                        fsExtra.createReadStream(options.destination).pipe(tar.extract(destWithoutTar))
                            .on('finish', function () {
                                fsExtra.remove(options.destination, function(err) {
                                    if(err) {
                                        return callback(new Error("Couldn't remove temporary destination (tar file generated when moving directories)"));
                                    }
                                    options.source = options.source.substr(0, options.source.length - 4);
                                    options.destination = options.destination.substr(0, options.destination.length - 4);
                                    return callback(null, sourceFileHash, dstFileHash);
                                })
                            });
                    }
                    else {
                        return callback(null, sourceFileHash, dstFileHash);
                    }
                });
            },
            function (sourceFileHash, dstFileHash, callback) {
                if (options.useLock) {
                    //Remove our lock, we are done!
                    var lockFilePath = options.destination + '.copy-lock';
                    lockFile.unlock(lockFilePath, function (err) {
                        if (err) {
                            return callback(err);
                        }
                        return callback(null, sourceFileHash, dstFileHash);
                    });
                }
                else {
                    return callback(null, sourceFileHash, dstFileHash);
                }
            }
        ],
        function (err, sourceFileHash, dstFileHash) {
            callback(err, {
                source: sourceFileHash,
                destination: dstFileHash
            });
        });
}

function getCopyOptions(options) {
    var resultOpts = {};
    //Overwrite
    resultOpts.clobber = (options.clobber !== undefined) ? options.clobber : false;

    resultOpts.srcHash = options.srcHash;
    resultOpts.hashMethod = options.hashMethod || 'md5';

    resultOpts.tmpSuffix = options.tmpSuffix || '.copy-tmp';
    resultOpts.useLock = (options.useLock !== undefined) ? options.useLock : true;

    resultOpts.stale = options.stale || 30000; //in ms
    resultOpts.wait = options.wait || 10000; //in ms
    return resultOpts;
}

function getFileHash(path, hashMethod, callback) {
    try {
        var cryptoHash = crypto.createHash(hashMethod);
    }
    catch (err) {
        //hashMethod not supported
        err.code = "BADHASH";
        return callback(err);
    }
    var hashHex = '';

    //Hash source
    var fileStream = fsExtra.ReadStream(path);
    fileStream.on('data', function (data) {
        cryptoHash.update(data);
    });
    fileStream.on('error', function (err) {
        return callback(err);
    });
    fileStream.on('end', function () {
        hashHex = cryptoHash.digest('hex');
        return callback(null, hashHex);
    });
}


module.exports = {
    copy: copy
};